kotlinx-coroutines-test/common/src/TestScope.kt

```diff
@@ -0,0 +1,219 @@
+ *   If you need to have a specific [CoroutineExceptionHandler], please pass it to [launch] on an already-created
+ *   [TestCoroutineScope] and share your use case at
+ *   [our issue tracker](https://github.com/Kotlin/kotlinx.coroutines/issues).
+ * * If [context] provides a [Job], that job is used as a parent for the new scope;
+ *   otherwise, a [CompletableJob] is created.
+ *
+ * @throws IllegalArgumentException if [context] has both [TestCoroutineScheduler] and a [TestDispatcher] linked to a
+ * different scheduler.
+ * @throws IllegalArgumentException if [context] has a [ContinuationInterceptor] that is not a [TestDispatcher].
+ * @throws IllegalArgumentException if [context] has an [CoroutineExceptionHandler] that is not an
+ * [UncaughtExceptionCaptor].
+ */
+@ExperimentalCoroutinesApi
+@Suppress("FunctionName")
+public fun TestScope(context: CoroutineContext = EmptyCoroutineContext): TestScope {
+    val ctxWithDispatcher = context.withDelaySkipping()
+    var scope: TestScopeImpl? = null
+    val exceptionHandler = when (ctxWithDispatcher[CoroutineExceptionHandler]) {
+        null -> CoroutineExceptionHandler { _, exception ->
+            scope!!.reportException(exception)
+        }
+        else -> throw IllegalArgumentException(
+            "A CoroutineExceptionHandler was passed to TestScope. " +
+                "Please pass it as an argument to a `launch` or `async` block on an already-created scope " +
+                "if uncaught exceptions require special treatment."
+        )
+    }
+    return TestScopeImpl(ctxWithDispatcher + exceptionHandler).also { scope = it }
+}
+
+/**
+ * Adds a [TestDispatcher] and a [TestCoroutineScheduler] to the context if there aren't any already.
+ *
+ * @throws IllegalArgumentException if both a [TestCoroutineScheduler] and a [TestDispatcher] are passed.
+ * @throws IllegalArgumentException if a [ContinuationInterceptor] is passed that is not a [TestDispatcher].
+ */
+internal fun CoroutineContext.withDelaySkipping(): CoroutineContext {
+    val dispatcher: TestDispatcher = when (val dispatcher = get(ContinuationInterceptor)) {
+        is TestDispatcher -> {
+            val ctxScheduler = get(TestCoroutineScheduler)
+            if (ctxScheduler != null) {
+                require(dispatcher.scheduler === ctxScheduler) {
+                    "Both a TestCoroutineScheduler $ctxScheduler and TestDispatcher $dispatcher linked to " +
+                        "another scheduler were passed."
+                }
+            }
+            dispatcher
+        }
+        null -> StandardTestDispatcher(get(TestCoroutineScheduler))
+        else -> throw IllegalArgumentException("Dispatcher must implement TestDispatcher: $dispatcher")
+    }
+    return this + dispatcher + dispatcher.scheduler
+}
+
+internal class TestScopeImpl(context: CoroutineContext) :
+    AbstractCoroutine<Unit>(context, initParentJob = true, active = true), TestScope {
```

Comment:
Is making `TestScopeImpl` an `AbstractCoroutine` a deliberate decision?

Because now we have [potentially unexpected] broken invariant, e.g.:
```
val root = Job()
val scope = TestScope(root)
println(scope.coroutineContext.job == root) // False
```

Attempt to attach a child to such scope is also visible side-effect of this change

Reply:
This is deliberate, though maybe I'm missing the downsides.

My reasoning: with `TestCoroutineScope`, we had this hierarchy:
```
/ manually-created TestCoroutineScope
| / the child coroutines spawned off before the test
| \ 
| / the coroutine that is a new scope
| | | child coroutines spawned inside the test
| \ cleanupTestCoroutines after runTest
\ cleanupTestCoroutines in afterTest
```
This structure ensured that there were no uncompleted child jobs.

`TestScope` does not provide a user-visible cleanup procedure, so if we want to detect that the child coroutines spawned off before the test are not leaked, we need to perform their cleanup as well at the end of the test:
```
/ manually-created TestScope (if it is not a coroutine)
| / the child coroutines spawned off before the test
| \ 
| / a coroutine TestScope
| | | child coroutines spawned inside the test
| \ ...
\ cleanup after runTest
```
Cleaning up the parent job doesn't look natural in this case.

With the proposed scheme, we have the following:

```
/ potential root job
| / TestScope
| | /
| | | the child coroutines spawned off before the test
| | | child coroutines spawned inside the test
| | \ 
\ \ cleanup after runTest
```

Reply:
That makes sense, thanks.

The downsides are kind of negligible -- the overall semantics is just not aligned with other scope factories. But it seems to be perfectly normal, taking into account all the limitations/machinery `TestScope` already has.